


# BloomTech - Hyperparameter Tuning Project

## Data Science Unit 4.2.3

Michael Luo

2023/03/01

# Gridsearch Hyperparameters

In the guided project, you learned how to use sklearn's `GridsearchCV` and `keras-tuner` libraries to tune the hyperparameters of a neural network model. For your module project you'll continue using these two libraries, however we are going to make things a little more interesting for you. 

Continue to use TensorFlow Keras & a sample of the [Quickdraw dataset](https://github.com/googlecreativelab/quickdraw-dataset) to build a sketch classification model. The dataset has been sampled to only 10 classes and 10000 observations per class. 



**Don't forget to switch to GPU on Colab!**

## 0.1 Imports and installs

In [ ]:
# native python libraries imports 
import math
from time import time
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

# sklearn imports 
import sklearn
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV

# keras imports 
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense
from tensorflow.keras.activations import relu, sigmoid
from tensorflow.keras.optimizers import Adam, SGD
from tensorflow.keras.utils import get_file

# required for compatibility between sklearn and keras
from tensorflow.keras.wrappers.scikit_learn import KerasClassifier

# install keras-tuner
!pip install keras-tuner
# from kerastuner.tuners import RandomSearch, BayesianOptimization, Sklearn ??
from kerastuner.tuners import RandomSearch, BayesianOptimization
from kerastuner.engine.hyperparameters import HyperParameters

## 0.2 Load quickdraw data

In [9]:
def load_quickdraw10():
    """
    Loads and normalizes (based on max pixel value) and partitions a subset of 
    the QuickDraw dataset into training and test sets.

    Returns
    -------
    X, y : shuffled train test split
    """
    
    URL_ = "https://github.com/michael-s-luo/Neural-Networks-DS4.2/blob/main/data/quickdraw10.npz?raw=true"
    
    path_to_zip = get_file('./quickdraw10.npz', origin=URL_, extract=False)

    data = np.load(path_to_zip)
    
    # normalize your image data
    max_pixel_value = 255
    X = data['arr_0']/max_pixel_value
    Y = data['arr_1']
        
    return train_test_split(X, Y, shuffle=True)

In [12]:
X_train, X_test, y_train, y_test = load_quickdraw10()

In [14]:
X_train.shape

(75000, 784)

In [15]:
y_train.shape

(75000,)

_____

# Experiment 1

## Tune Hyperperameters using Enhanced GridsearchCV 

We are going to use GridsearchCV again to tune a deep learning model however we are going to add some additional functionality to our gridsearch. 

Specifically, we are going to automate the generation of how many nodes to use in a layer and how many layers to use in a model! 

By the way, yes, there is a function within a function. Try to not let that bother you. An alternative to this would be to create a class. If you're up for the challenge give it a shot. However, consider this a stretch goal that you come back to after you finish going through this assignment. 


### Objective 

The objective of this experiment is to show you how to automate the generation of layers and layer nodes for the purposes of gridsearch. <br>
Up until now, we've been manually selecting the number of layers and layer nodes.

In [65]:
# Function to create model, required for KerasClassifier
def create_model(n_layers,  first_layer_nodes, last_layer_nodes, act_funct ="relu", negative_node_incrementation=True):
    """"
    Returns a compiled keras model 
    
    Parameters
    ----------
    n_layers: int 
        number of hidden layers in model 
        To be clear, this excludes the input and output layer.
        
    first_layer_nodes: int
        Number of nodes in the first hidden layer 

    last_layer_nodes: int
        Number of nodes in the last hidden layer (this is the layer just prior to the output layer)
        
     act_funct: string 
         Name of activation function to use in hidden layers (this excludes the output layer)
        
    Returns
    -------
    model: keras object 
    """
    
    def gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation=True):
        """
        Generates and returns the number of nodes in each hidden layer. 
        To be clear, this excludes the input and output layer. 

        Note
        ----
        Number of nodes in each layer is linearly incremented. 
        For example, gen_layer_nodes(5, 500, 100) will generate [500, 400, 300, 200, 100]

        Parameters
        ----------
        n_layers: int
            Number of hidden layers
            This values should be 2 or greater 

        first_layer_nodes: int

        last_layer_nodes: int

        Returns
        -------
        layers: list of ints
            Contains number of nodes for each layer 
        """

        # throws an error if n_layers is less than 2 
        assert n_layers >= 2, "n_layers must be 2 or greater"

        layers = []

        # PROTIP: IF YOU WANT THE NODE INCREMENTATION TO BE SPACED DIFFERENTLY
        # THEN YOU'LL NEED TO CHANGE THE WAY THAT IT'S CALCULATED 
        # when set to True number of nodes are decreased for subsequent layers 
        # NOTE: the order of the number of nodes doesn't matter
        if negative_node_incrementation:
            # subtract this amount from previous layer's nodes in order to increment towards smaller numbers 
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
            
        # when set to False number of nodes are increased for subsequent layers
        else:
            # add this amount from previous layer's nodes in order to increment towards larger numbers 
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)

        nodes = first_layer_nodes

        for i in range(1, n_layers+1):

            layers.append(nodes)

            # increment nodes for next layer 
            nodes = nodes + nodes_increment

        return np.round(layers)
    
    # create model
    model = Sequential()
    
    n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
    
    for i in range(0, n_layers):
        if i==0:
            model.add(Dense(first_layer_nodes, input_dim=X_train.shape[1], activation=act_funct))
        else:
            model.add(Dense(n_nodes[i], activation=act_funct))
            
    # output layer 
    model.add(Dense(10, # 10 unit/neurons in output layer because we have 10 possible labels to predict  
                    activation='softmax')) # use softmax for a label set greater than 2            
    
    # Compile model
    model.compile(loss='sparse_categorical_crossentropy', 
                  optimizer='adam', # adam is a good default optimizer 
                  metrics=['accuracy'])
    
    # do not include model.fit() inside the create_model function
    # KerasClassifier is expecting a complied model 
    return model


## 1.1 Explore `create_model`



The helper function `gen_layer_nodes()` which is contained inside `create_model()` <br>
returns a list containing the number of nodes for each successive layer.<br>

Let's check that `gen_layer_nodes()` behaves as expected. <br>
In other words, we'll perform a **Unit Test!**

In [63]:
    def gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation=True):
        """
        Generates and returns the number of nodes in each hidden layer. 
        To be clear, this excludes the input and output layer. 

        Note
        ----
        Number of nodes in each layer is linearly incremented. 
        For example, gen_layer_nodes(5, 500, 100) will generate [500, 400, 300, 200, 100]

        Parameters
        ----------
        n_layers: int
            Number of hidden layers
            This values should be 2 or greater 

        first_layer_nodes: int

        last_layer_nodes: int

        Returns
        -------
        layers: list of ints
            Contains number of nodes for each layer 
        """

        # throws an error if n_layers is less than 2 
        assert n_layers >= 2, "n_layers must be 2 or greater"

        layers = []

        # PROTIP: IF YOU WANT THE NODE INCREMENTATION TO BE SPACED DIFFERENTLY
        # THEN YOU'LL NEED TO CHANGE THE WAY THAT IT'S CALCULATED - HAVE FUN!
        # when set to True number of nodes are decreased for subsequent layers 
        # NOTE: the order of the number of nodes doesn't matter
        if negative_node_incrementation:
            # subtract this amount from previous layer's nodes in order to increment towards smaller numbers 
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
            #print(f'nodes increment = {nodes_increment}')
            
        # when set to False number of nodes are increased for subsequent layers
        else:
            # add this amount from previous layer's nodes in order to increment towards larger numbers 
            nodes_increment = (last_layer_nodes - first_layer_nodes)/ (n_layers-1)
            #print(f'nodes increment = {nodes_increment}')

        nodes = first_layer_nodes

        for i in range(1, n_layers+1):

            layers.append(nodes)

            # increment nodes for next layer 
            nodes = nodes + nodes_increment

        return np.round(layers)




### `negative_node_incrementation = True`
For this case we want the number of nodes to _decrease_ by a constant number for successive layers. <br>So `first_layer_nodes` must be _larger_ than `last_layer_nodes` 

In [64]:
# This function was provided by the BloomTech curriculum.
# negative_node_incrementation has no effect on the n_nodes outcome (which is a list of number of nodes per layer in succession)
  # The function always subtracts first_layer_nodes from last_layer_nodes; already outputs correct directional increment

# n_layers = 10
# first_layer_nodes = 100
# last_layer_nodes = 500
# negative_node_incrementation = True
# n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
# print(f'Number of nodes in successive layers: {n_nodes}')

Number of nodes in successive layers: [100. 144. 189. 233. 278. 322. 367. 411. 456. 500.]


### `negative_node_incrementation = False`
For this case we want the number of nodes to _increase_ by a constant number for successive layers. <br>So `first_layer_nodes` must be _smaller_ than `last_layer_nodes` 

In [26]:
# This function was provided by the BloomTech curriculum.
# negative_node_incrementation has no effect on the n_nodes outcome (which is a list of number of nodes per layer in succession)
  # The function always subtracts first_layer_nodes from last_layer_nodes; already outputs correct directional increment


# n_layers = 5
# first_layer_nodes = 100
# last_layer_nodes = 500
# negative_node_incrementation = False
# n_nodes = gen_layer_nodes(n_layers, first_layer_nodes, last_layer_nodes, negative_node_incrementation)
# print(f'Number of nodes in successive layers: {n_nodes}')

Number of nodes in successive layers: [100, 200, 300, 400, 500]


### OK, the Unit Test is passed!

### Let's build a few models<br> 
in order to understand how `create_model()` works in practice. 

### Build a model, setting `negative_node_incrementation = True` 

Use `create_model` to build a model. 

- Set `n_layers = 10` 
- Set `first_layer_nodes = 500`
- Set `last_layer_nodes = 100`
- Set `act_funct = "relu"`


In [43]:
# use create_model to create a model 

# YOUR CODE HERE
model = create_model(n_layers = 10, first_layer_nodes = 500, last_layer_nodes = 100)

In [49]:
model.get_config()

{'name': 'sequential_4',
 'layers': [{'class_name': 'InputLayer',
   'config': {'batch_input_shape': (None, 784),
    'dtype': 'float32',
    'sparse': False,
    'ragged': False,
    'name': 'dense_26_input'}},
  {'class_name': 'Dense',
   'config': {'name': 'dense_26',
    'trainable': True,
    'dtype': 'float32',
    'batch_input_shape': (None, 784),
    'units': 500,
    'activation': 'relu',
    'use_bias': True,
    'kernel_initializer': {'class_name': 'GlorotUniform',
     'config': {'seed': None}},
    'bias_initializer': {'class_name': 'Zeros', 'config': {}},
    'kernel_regularizer': None,
    'bias_regularizer': None,
    'activity_regularizer': None,
    'kernel_constraint': None,
    'bias_constraint': None}},
  {'class_name': 'Dense',
   'config': {'name': 'dense_27',
    'trainable': True,
    'dtype': 'float32',
    'units': 456,
    'activation': 'relu',
    'use_bias': True,
    'kernel_initializer': {'class_name': 'GlorotUniform',
     'config': {'seed': None}},
   

In [45]:
# run model.summary() and make sure that you understand the model architecture that you just built 
# Notice in the model summary how the number of nodes have been linearly incremented in decreasing values. 
model.summary()

Model: "sequential_4"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_26 (Dense)            (None, 500)               392500    
                                                                 
 dense_27 (Dense)            (None, 456)               228456    
                                                                 
 dense_28 (Dense)            (None, 412)               188284    
                                                                 
 dense_29 (Dense)            (None, 367)               151571    
                                                                 
 dense_30 (Dense)            (None, 323)               118864    
                                                                 
 dense_31 (Dense)            (None, 278)               90072     
                                                                 
 dense_32 (Dense)            (None, 234)              

### Build a model, setting `negative_node_incrementation = False` 

Use `create_model` to build a model. 

- Set `n_layers = 10` 
- Set `first_layer_nodes = 100`
- Set `last_layer_nodes = 500`
- Set `act_funct = "relu"`

In [66]:
# use create_model to create a model 

model = create_model(n_layers=10, first_layer_nodes=100, last_layer_nodes=500)

In [67]:
# run model.summary() and make sure that you understand the model architecture that you just built 
# Notice in the model summary how the number of nodes have been linearly incremented in decreasing values. 
model.summary()

Model: "sequential_7"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
 dense_59 (Dense)            (None, 100)               78500     
                                                                 
 dense_60 (Dense)            (None, 144)               14544     
                                                                 
 dense_61 (Dense)            (None, 189)               27405     
                                                                 
 dense_62 (Dense)            (None, 233)               44270     
                                                                 
 dense_63 (Dense)            (None, 278)               65052     
                                                                 
 dense_64 (Dense)            (None, 322)               89838     
                                                                 
 dense_65 (Dense)            (None, 367)              

## 1.2 Create a grid search using `sklearn`

### Hyperparameter search

In [68]:
# define the grid search parameters
param_grid = {'n_layers': [2, 3],
              'epochs': [3], 
              "first_layer_nodes": [500, 300],
              "last_layer_nodes": [100, 50]
             }

In [70]:
model = KerasClassifier(build_fn=create_model)

<ipython-input-70-d5ce3444c54d>:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  model = KerasClassifier(build_fn=create_model)


In [71]:
%%time
# Create Grid Search
grid = GridSearchCV(estimator=model, 
                    param_grid=param_grid, 
                    n_jobs=-2, 
                    verbose=2, 
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 8 candidates, totalling 24 fits
Epoch 1/3
1563/1563 [==============================] - 10s 3ms/step - loss: 0.6429 - accuracy: 0.8060
Epoch 2/3
1563/1563 [==============================] - 5s 3ms/step - loss: 0.4345 - accuracy: 0.8690
Epoch 3/3
782/782 [==============================] - 2s 2ms/step - loss: 0.4982 - accuracy: 0.8552
[CV] END epochs=3, first_layer_nodes=500, last_layer_nodes=100, n_layers=2; total time=  24.4s
Epoch 1/3
1563/1563 [==============================] - 7s 4ms/step - loss: 0.6454 - accuracy: 0.8068
Epoch 2/3
1563/1563 [==============================] - 5s 3ms/step - loss: 0.4344 - accuracy: 0.8684
Epoch 3/3
782/782 [==============================] - 2s 3ms/step - loss: 0.4479 - accuracy: 0.8674
[CV] END epochs=3, first_layer_nodes=500, last_layer_nodes=100, n_layers=2; total time=  24.5s
Epoch 1/3
1563/1563 [==============================] - 7s 4ms/step - loss: 0.6437 - accuracy: 0.8050
Epoch 2/3
1563/1563 [=========================

In [74]:


best_model = grid_result.best_estimator_
best_model.get_params()

{'epochs': 3,
 'first_layer_nodes': 500,
 'last_layer_nodes': 50,
 'n_layers': 2,
 'build_fn': <function __main__.create_model(n_layers, first_layer_nodes, last_layer_nodes, act_funct='relu', negative_node_incrementation=True)>}

In [88]:
layers = grid_result.cv_results_['param_n_layers']
first_layer_nodes = grid_result.cv_results_['param_first_layer_nodes']
last_layer_nodes = grid_result.cv_results_['param_last_layer_nodes']
cv_accuracies = grid_result.cv_results_['mean_test_score']
ranking = grid_result.cv_results_['rank_test_score']

pd.DataFrame(data={"Number of Layers": layers, 
                   "1st Layer Nodes": first_layer_nodes, 
                   "Last Layer Nodes": last_layer_nodes, 
                   "Mean CV Accuracy": cv_accuracies, 
                   "Rank":ranking}).sort_values(by=["Rank"])

Number of Layers 1st Layer Nodes Last Layer Nodes  Mean CV Accuracy  Rank
2                2             500               50          0.869267     1
1                3             500              100          0.866800     2
7                3             300               50          0.865827     3
4                2             300              100          0.864427     4
3                3             500               50          0.862747     5
0                2             500              100          0.862613     6
5                3             300              100          0.861373     7
6                2             300               50          0.857053     8

In [86]:
print

grid_result.cv_results_.keys()

dict_keys(['mean_fit_time', 'std_fit_time', 'mean_score_time', 'std_score_time', 'param_epochs', 'param_first_layer_nodes', 'param_last_layer_nodes', 'param_n_layers', 'params', 'split0_test_score', 'split1_test_score', 'split2_test_score', 'mean_test_score', 'std_test_score', 'rank_test_score'])

Ok, now that we've played around a bit with  `create_model`, let's build a  simpler model that we'll use to run gridsearches. 

-----

# Experiment 2: Run the Gridsearch Algorithms 

In this section, we are going to use the same model and dataset in order to benchmark 3 different gridsearch approaches: 

- Gridsearch
- Random Search
- Bayesian Optimization. 


Our goal in this experiment is two-fold. We want to see which appraoch 

- Scores the highest accuracy
- Has the shortest run time 

We want to see how these 3 gridsearch approaches handle these trade-offs and to give you a sense of those trades offs.

### Trade-offs

`Gridsearch` will train a model on every single unique hyperparameter combination, this guarantees that you'll get the highest possible accuracy from your parameter set but your gridsearch might have a very long run-time. 

`Random Search` will randomly sample from your parameter set which, depending on how many samples, the run-time might be significantly cut down but you might or might not sample the parameters that correspond to the heightest possible accuracies. 

`Bayesian Optimization` has a bit of intelligence built into it's search algorithm but you do need to manually select some parameters which may greatly influence the model learning outcomes. 

-------
### Build our model

In [89]:
# because gridsearching can take a lot of time and we are bench marking 3 different approaches
# let's build a simple model to minimize run time 

def build_model(hp):
    
    """
    Returns a complied keras model ready for keras-tuner gridsearch algorithms 
    """
    
    model = Sequential()
    
    # hidden layer
    model.add(Dense(units=hp.get('units'),activation=hp.get("activation")))
    
    # output layer
    model.add(Dense(10, activation='softmax'))
    
    model.compile(
        optimizer=Adam(hp.get('learning_rate')),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    
    return model
  

In [102]:
# build out our hyperparameter dictionary 
hp = HyperParameters()
hp.Int('units', min_value=32, max_value=512, step=32)
hp.Choice('learning_rate',values=[1e-1, 1e-2, 1e-3])
hp.Choice('activation',values=["relu", "sigmoid"])

'relu'

---------
## 2.1 Gridsearch Optimization


### Populate a `sklearn` compatible parameter dictionary

In [104]:
# build out our hyperparameter dictionary 
hyper_parameters = {
    # BUG Fix: cast array as list otherwise GridSearchCV will throw error
    "units": np.arange(32, 512, 32).tolist(),
    "learning_rate": [1e-1, 1e-2, 1e-3],
    "activation":["relu", "sigmoid"]
}

In [ ]:
hyper_parameters

### Build a `sklearn` compatible model function

In [112]:
def build_model(units, learning_rate, activation):
    
    """
    Returns a complie keras model ready for keras-tuner gridsearch algorithms 
    """
    
    model = Sequential()
    
    # hidden layer
    model.add(Dense(units, activation=activation))
    
    # output layer
    model.add(Dense(10, activation='softmax'))
    
    model.compile(
        optimizer=Adam(learning_rate),
        loss='sparse_categorical_crossentropy',
        metrics=['accuracy'])
    
    return model

### Apply the "wrapper" to make the model compatible with `sklearn`

In [113]:
model = KerasClassifier(build_fn = build_model)

<ipython-input-113-a93dc876c14b>:1: DeprecationWarning: KerasClassifier is deprecated, use Sci-Keras (https://github.com/adriangb/scikeras) instead. See https://www.adriangb.com/scikeras/stable/migration.html for help migrating.
  model = KerasClassifier(build_fn = build_model)


In [114]:
# save start time 
start = time()

# Create Grid Search
grid = GridSearchCV(estimator=model, 
                    param_grid=hyper_parameters, 
                    n_jobs=-2, 
                    verbose=1, 
                    cv=3)

grid_result = grid.fit(X_train, y_train)

# save end time 
end = time()

# Report Results
print(f"Best: {grid_result.best_score_} using {grid_result.best_params_}")

means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']

for mean, stdev, param in zip(means, stds, params):
    print(f"Means: {mean}, Stdev: {stdev} with: {param}")

Fitting 3 folds for each of 90 candidates, totalling 270 fits
2344/2344 [==============================] - 10s 4ms/step - loss: 0.6177 - accuracy: 0.8157
Best: 0.8427866498629252 using {'activation': 'relu', 'learning_rate': 0.001, 'units': 416}
Means: 0.26499999066193897, Stdev: 0.024814985818366556 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 32}
Means: 0.28411999344825745, Stdev: 0.013614147526136825 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 64}
Means: 0.24505333105723062, Stdev: 0.06177392651802086 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 96}
Means: 0.24634667237599692, Stdev: 0.028939704736259234 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 128}
Means: 0.32603999972343445, Stdev: 0.0005256139076642272 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 160}
Means: 0.2817866653203964, Stdev: 0.047415236936773096 with: {'activation': 'relu', 'learning_rate': 0.1, 'units': 192}
Means: 0.2595600038766861, Stdev

In [115]:
# total run time 
total_run_time_in_miniutes = (end - start)/60
total_run_time_in_miniutes

52.86374278863271

In [116]:
# Best parameter combination
grid_result.best_params_

{'activation': 'relu', 'learning_rate': 0.001, 'units': 416}

In [117]:
# because all other optimization approaches are reporting test set score
# let's calculate the test set score in this case 
best_model = grid_result.best_estimator_

test_acc = best_model.score(X_test, y_test)

782/782 [==============================] - 3s 3ms/step - loss: 0.4933 - accuracy: 0.8539


In [119]:
test_acc, grid_result.best_score_

(0.8539199829101562, 0.8427866498629252)

 ### Results
 
Identify and write the the best performing hyperparameter combination and model score. 
 
 

The grid search spanned a total of ~53 minutes!

Best hyperparameters:

'activation': 'relu'

'learning_rate': 0.001

'units': 416

This model resulted in validation accuracy of 0.8428 and testing accuracy of 0.8539. These are relatively lower than those of models found by keras-tuner algorithms; each model in the grid-search was only trained for 1 epoch. Validation and testing accuracy are expected to be higher.



------
## 2.2 Random Search with `keras-tuner`

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `RandomSearch` tuner.

In [ ]:
# how many unique hyperparameter combinations do we have? 
# HINT: take the product of the number of possible values for each hyperparameter 

n_unique_hparam_combos = 90 # does not include 512 units

In [ ]:
# how many of these do we want to randomly sample?
# let's pick 25% of n_unique_hparam_combos param combos to sample
# 90 unique combos -> ! 23 for random search

fraction_to_sample = 0.25
n_param_combos_to_sample = np.round(n_unique_hparam_combos * fraction_to_sample)


### Instantiate a `RandomSearch()` object for your grid search

In [ ]:
random_tuner = RandomSearch(
            build_model,
            objective='val_accuracy', # this does not use cross-validation 
            max_trials=n_param_combos_to_sample, # number of times to sample the parameter set and build a model 
            seed=1234,
            hyperparameters=hp, # pass in our hyperparameter dictionary
            directory='./keras-tuner-trial',
            project_name='random_search')

In [ ]:
 # take note of Total elapsed time in print out -- took ~10 minutes without GPU
random_tuner.search(X_train, y_train,
                    epochs=3,
                    validation_data=(X_test, y_test)) # shouldn't validate on testing data; choosing the best RandomSearch model here 'leaks' information into our model
                                                      # need to partition from training data for validation

Trial 20 Complete [00h 00m 34s]
val_accuracy: 0.8131999969482422

Best val_accuracy So Far: 0.8730000257492065
Total elapsed time: 00h 10m 09s
INFO:tensorflow:Oracle triggered exit


In [ ]:
# identify the best score and hyperparamter (should be at the top since scores are ranked)
random_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/random_search
Showing 10 best trials
Objective(name='val_accuracy', direction='max')
Trial summary
Hyperparameters:
units: 384
learning_rate: 0.001
activation: relu
Score: 0.8730000257492065
Trial summary
Hyperparameters:
units: 352
learning_rate: 0.001
activation: relu
Score: 0.8729199767112732
Trial summary
Hyperparameters:
units: 448
learning_rate: 0.001
activation: sigmoid
Score: 0.8636400103569031
Trial summary
Hyperparameters:
units: 480
learning_rate: 0.001
activation: sigmoid
Score: 0.8621199727058411
Trial summary
Hyperparameters:
units: 416
learning_rate: 0.001
activation: sigmoid
Score: 0.8591200113296509
Trial summary
Hyperparameters:
units: 288
learning_rate: 0.001
activation: sigmoid
Score: 0.8572400212287903
Trial summary
Hyperparameters:
units: 224
learning_rate: 0.001
activation: sigmoid
Score: 0.8525599837303162
Trial summary
Hyperparameters:
units: 320
learning_rate: 0.01
activation: sigmoid
Score: 0.8378400206565857
Tri

 ### Results
 
Identify and write the the best performing hyperparameter combination and model score. 
Note that because this is Random Search, multiple runs might have slighly different outcomes. 
 
 

From the random_tuner.results_summary():

Units: 384

Learning_rate: 0.001

A
ctivation: relu

------
## 2.3 Bayesian Optimization with `keras-tuner`

![](https://upload.wikimedia.org/wikipedia/commons/0/02/GpParBayesAnimationSmall.gif)

Be sure to check out the [**docs for Keras-Tuner**](https://keras-team.github.io/keras-tuner/documentation/tuners/). Here you can read about the input parameters for the `BayesianOptimization` tuner.

Pay special attention to these `BayesianOptimization` parameters: `num_initial_points` and `beta`. 

`num_initial_points`: 

Number of randomly selected hyperparameter combinations to try before applying bayesian probability to determine likelihood of which param combo to try next based on expected improvement




`beta`: 

Larger values means more willing to explore new hyperparameter combinations (analogous to searching for the global minimum in Gradient Descent), smaller values means that it is less willing to try new hyperparameter combinations (analogous to getting stuck in a local minimum in Gradient Descent). 

As a start, error on the side of larger values. What defines a small or large value you ask? That question would pull us into the mathematical intricacies of Bayesian Optimization and Gaussian Processes. For simplicity, notice that the default value is 2.6 and work from there. 

In [ ]:
# we know that 23 samples is about 25% of 90 possible hyper-parameter combos
# let's set up a run with the same parameters we used for RandomSearch() so the comparison will be apples-to-apples
# feel free to play with any of these numbers later
max_trials=23
num_initial_points=5
beta=5.0

#### Instantiate a `BayesianOptimization()` object for your grid search

In [ ]:
bayesian_tuner = BayesianOptimization(
                    build_model,
                    objective='val_accuracy',
                    max_trials=max_trials,
                    hyperparameters=hp, # pass in our hyperparameter dictionary
                    num_initial_points=num_initial_points, 
                    beta=beta, 
                    seed=1234,
                    directory='./keras-tuner-trial',
                    project_name='bayesian_optimization_4')

INFO:tensorflow:Reloading Oracle from existing project ./keras-tuner-trial/bayesian_optimization_4/oracle.json
INFO:tensorflow:Reloading Tuner from ./keras-tuner-trial/bayesian_optimization_4/tuner0.json


In [ ]:
bayesian_tuner.search(X_train, y_train,
               epochs=3,
               validation_data=(X_test, y_test))

Trial 9 Complete [00h 00m 21s]
val_accuracy: 0.8266800045967102

Best val_accuracy So Far: 0.875
Total elapsed time: 00h 03m 05s
INFO:tensorflow:Oracle triggered exit


In [ ]:
bayesian_tuner.results_summary()

Results summary
Results in ./keras-tuner-trial/bayesian_optimization_4
Showing 10 best trials
Objective(name='val_accuracy', direction='max')
Trial summary
Hyperparameters:
units: 512
learning_rate: 0.001
activation: relu
Score: 0.875
Trial summary
Hyperparameters:
units: 352
learning_rate: 0.001
activation: relu
Score: 0.8678799867630005
Trial summary
Hyperparameters:
units: 480
learning_rate: 0.001
activation: sigmoid
Score: 0.860319972038269
Trial summary
Hyperparameters:
units: 256
learning_rate: 0.001
activation: sigmoid
Score: 0.8562399744987488
Trial summary
Hyperparameters:
units: 352
learning_rate: 0.01
activation: sigmoid
Score: 0.8393200039863586
Trial summary
Hyperparameters:
units: 512
learning_rate: 0.01
activation: sigmoid
Score: 0.8347200155258179
Trial summary
Hyperparameters:
units: 32
learning_rate: 0.001
activation: relu
Score: 0.8316400051116943
Trial summary
Hyperparameters:
units: 32
learning_rate: 0.001
activation: relu
Score: 0.83024001121521
Trial summary
Hype

 ### Results
 
Identify and write the the best performing hyperparameter combination and model score. 
Note that because this is  Bayesian Optimization, multiple runs might have slighly different outcomes. 
 
 

From bayesian_tuner.results_summary():

Best Hyperparameter Combination:

units: 512

learning_rate: 0.001

activation: relu

This model resulted in a validation accuracy of 0.875

We should point out that Gridsearch split the training set internally and performed 3-fold cross-validation whereas keras-tuner required a validation/test set. This means that the keras-tuner algorithms were using one validation set and our skearn GridSearchCV was using a different set - so this isn't a perfectly exact 1-to-1 comparision but it'll have to do. 

In order to compensate for this, we did score the best model on the same test set that keras-tuner used. Because the best-performing model from keras-tuner used our QuickDraw testing holdout data for validation, the test holdout is no longer valid as "new, unseen data". Information has leaked into our modeling process; namely, the best model was chosen based on the testing dataset. Ideally, keras-tuner should support cross-validation so that we do not need to create a separate validation holdout from our training data. 

_______

# Conclusion

The spirit of this experiment is to expose you to the idea of benchmarking and comparing the trade-offs of various gridsearch approaches. 

Even if we did find a way to pass in the original test set into GridSearchCV, we can see that both Random Search and Bayesian Optimization are arguably better alternatives to a brute force grid search when we consider the trade-offs of run time and locating the best performing model. 